File: test/python/element_pairing.py
from helper import *

class TestElementPairing(unittest.TestCase):
    def setUp(self):
        self.pairing = Pairing.init_from_param_str(TYPE_A_PARAM)
        self.e1 = Element(self.pairing, ElementType_G1)
        self.e2 = Element(self.pairing, ElementType_G2)
        self.z = Element(self.pairing, ElementType_Zr)
        self.e1.random()
        self.e2.random()
        self.z.random()

    def test_e(self):
        with self.assertRaises(RuntimeError):
            e(Element(), Element())
        et = e(self.e1, self.e2)
        self.assertEqual(et.type(), ElementType_GT)
        self.assertEqual(et ** self.z, e(self.e1 ** self.z, self.e2))

    def test_e_prod(self):
        with self.assertRaises(ValueError):
            e_prod([], [])
        with self.assertRaises(ValueError):
            e_prod([self.e1, self.e1], [self.e2])
        with self.assertRaises(RuntimeError):
            e_prod([self.e1], [Element()])
        et1 = e_prod([self.e1, self.e1], [self.e2, self.e2])
        et2 = e(self.e1, self.e2)
        self.assertEqual(et1, et2 * et2)

    def test_pairing_fixed_g1(self):
        with self.assertRaises(RuntimeError):
            FixedG1Pairing(Element())
        with self.assertRaises(ValueError):
            FixedG1Pairing(self.z)
        p = FixedG1Pairing(self.e1)
        with self.assertRaises(RuntimeError):
            p.apply(Element())
        self.assertEqual(p.apply(self.e2), e(self.e1, self.e2))

if __name__ == '__main__':
    unittest.main()

File: test/ruby/element_pairing.rb
require_relative "helper"

class TestElementPairing < TestCase
  def setup
    @pairing = Pairing.init_from_param_str(TYPE_A_PARAM)
    @e1 = Element.new(@pairing, ElementType_G1)
    @e2 = Element.new(@pairing, ElementType_G2)
    @z = Element.new(@pairing, ElementType_Zr)
    @e1.random
    @e2.random
    @z.random
  end

  def test_e
    assert_raise RuntimeError do
      e(Element.new, Element.new)
    end
    et = e(@e1, @e2)
    assert_equal ElementType_GT, et.type
    assert_equal et ** @z, e(@e1 ** @z, @e2)
  end

  def test_e_prod
    assert_raise ArgumentError do
      e_prod [], []
    end
    assert_raise ArgumentError do
      e_prod [@e1, @e1], [@e2]
    end
    assert_raise RuntimeError do
      e_prod([@e1], [Element.new])
    end
    et1 = e_prod [@e1, @e1], [@e2, @e2]
    et2 = e(@e1, @e2)
    assert_equal et2 * et2, et1
  end

  def test_pairing_fixed_g1
    assert_raise RuntimeError do
      FixedG1Pairing.new Element.new
    end
    assert_raise ArgumentError do
      FixedG1Pairing.new @zr
    end
    p = FixedG1Pairing.new @e1
    assert_equal @e1, p.get_g1
    assert_raise RuntimeError do
      p.apply Element.new
    end
    assert_equal e(@e1, @e2), p.apply(@e2)
  end
end
